experiments/yarn/src/main/scala/org/apache/gearpump/experiments/yarn/ResouceManagerClientActor.scala

```diff
@@ -1,44 +1,47 @@
 package org.apache.gearpump.experiments.yarn
 
-import org.apache.gearpump.experiments.yarn.Actions.AMStatusMessage
-import org.apache.gearpump.experiments.yarn.Actions.ContainerRequestMessage
-import org.apache.gearpump.experiments.yarn.Actions.RegisterAMMessage
-import org.apache.gearpump.experiments.yarn.master.ResourceManagerCallbackHandler
-import org.apache.gearpump.experiments.yarn.master.YarnApplicationMaster.TIME_INTERVAL
+import akka.actor.{Actor, actorRef2Scala}
+import org.apache.gearpump.experiments.yarn.master.{AmActorProtocol, ResourceManagerCallbackHandler, YarnApplicationMaster}
 import org.apache.gearpump.util.LogUtil
-import org.apache.hadoop.yarn.api.records.Priority
-import org.apache.hadoop.yarn.api.records.Resource
+import org.apache.hadoop.yarn.api.records.{Priority, Resource}
 import org.apache.hadoop.yarn.client.api.AMRMClient.ContainerRequest
 import org.apache.hadoop.yarn.client.api.async.AMRMClientAsync
 import org.apache.hadoop.yarn.conf.YarnConfiguration
 import org.apache.hadoop.yarn.util.Records
-import akka.actor.Actor
-import akka.actor.ActorRef
-import akka.actor.actorRef2Scala
-import org.apache.gearpump.experiments.yarn.master.YarnApplicationMaster
 
-class ResourceManagerClientActor(yarnConf: YarnConfiguration, yarnAM: ActorRef) extends Actor {
+import scala.util.{Failure, Success, Try}
+
+class ResourceManagerClientActor(yarnConf: YarnConfiguration, clientFactory : AMRMClientAsyncFactory[ContainerRequest]) extends Actor {
+
+  import AmActorProtocol._
+
   val LOG = LogUtil.getLogger(getClass)
   var client: AMRMClientAsync[ContainerRequest] = _
   
   override def receive: Receive = {
     case rmCallbackHandler: ResourceManagerCallbackHandler =>
       LOG.info("Received RMCallbackHandler")
-      client = start(rmCallbackHandler)
+      Try({
+        client = start(rmCallbackHandler)
+        sender ! AMRMClientAsyncStartup(Try(true))
+      }).failed.map(throwable => {
+        sender ! AMRMClientAsyncStartup(Failure(throwable))
+      })
     case containerRequest: ContainerRequestMessage =>
       LOG.info("Received ContainerRequestMessage")
       client.addContainerRequest(createContainerRequest(containerRequest))
     case amAttr: RegisterAMMessage =>
       LOG.info(s"Received RegisterAMMessage! ${amAttr.appHostName}:${amAttr.appHostPort}${amAttr.appTrackingUrl}")
       val response = client.registerApplicationMaster(amAttr.appHostName, amAttr.appHostPort, amAttr.appTrackingUrl)
-      LOG.info("got response : " + response)
-      yarnAM ! response
-    case amStatus: AMStatusMessage =>
+      LOG.info("sending response : " + response)
+      sender ! RegisterAppMasterResponse(response=response)
+    case amStatus:
+      AMStatusMessage =>
       LOG.info("Received AMStatusMessage")
       client.unregisterApplicationMaster(amStatus.appStatus, amStatus.appMessage, amStatus.appTrackingUrl)
```

Comment:
It is confusing to me why it do unregisterApplicationMaster when it received message like AmStatusMessage?

I don't find comments about AMStatusMessage, is it a normal status of Application Master? or exit status?